ui/radar/radar.js

```diff
@@ -0,0 +1,235 @@
+    if (monster['rank'] in options.RankOptions)
+      options = Object.assign({}, this.options, options.RankOptions[monster['rank']]);
+    let tr = monster['dom'];
+    // calculate rotation based on facing
+    let playerVector = new Point2D(e.detail.pos.x, -e.detail.pos.y);
+    let targetVector = monster['pos'];
+    let deltaVector = new Point2D(targetVector.x - playerVector.x, targetVector.y - playerVector.y);
+    if (tr) {
+      tr.childNodes[1].innerHTML = monster['rank'] + '&nbsp;&nbsp;&nbsp;&nbsp;' + monster['name'];
+      if (Math.abs(e.detail.pos.z - monster['pos_z']) > 5)
+        tr.childNodes[1].innerHTML += '&nbsp;&nbsp;' + (e.detail.pos.z < monster['pos_z']? '↑' : '↓');
+      tr.childNodes[1].innerHTML += '<br>' + length(deltaVector).toFixed(2) + 'm';
+      if (Date.now() / 1000 <= monster['battle_time'] + 60) {
+        tr.childNodes[1].innerHTML += ' ' + (monster['current_hp'] * 100 /
+          monster['hp']).toFixed(2) + '%';
+      }
+      if (options.Puller) {
+        if (monster['puller'])
+          tr.childNodes[1].innerHTML += '&nbsp;&nbsp;' + monster['puller'];
+      }
+    }
+    if (options.DetectionRange > 0 && length(deltaVector) > options.DetectionRange)
+      monster['dom'].setAttribute('class', 'hide');
+    else
+      monster['dom'].setAttribute('class', '');
+    let deltaTheta = Math.acos(theta(deltaVector, new Point2D(1, 0)));
+    if (deltaVector.y < 0)
+      deltaTheta = -deltaTheta;
+    deltaTheta += Math.PI - e.detail.rotation;
+    let angle = -deltaTheta*180/Math.PI;
+    let arrow_id = 'arrow-' + monster['id'];
+    let arrow = document.getElementById(arrow_id);
+    arrow.style.transform='rotate('+angle+'deg)';
+  }
+
+  RemoveMonster(mob_name) {
+    if (mob_name in this.targetMonsters) {
+      this.targetMonsters[mob_name]['dom'].remove();
+      delete this.targetMonsters[mob_name];
+    }
+  }
+
+  OnLogEvent(e) {
+    let lang = this.options.Language;
+    for (let i = 0; i < e.detail.logs.length; i++) {
+      // add new combatant
+      let matches = e.detail.logs[i].match(Regexes.addedCombatantFull({}));
+      if (matches) {
+        let monster = this.nameToMonster[matches.groups.name];
+        if (monster)
+          this.AddMonster(e.detail.logs[i], monster);
+      }
+      // network ability
+      matches = e.detail.logs[i].match(Regexes.abilityFull({}));
+      if (matches) {
+        let monster = this.targetMonsters[matches.groups.target];
+        if (monster)
+          this.UpdateMonsterPuller(e.detail.logs[i], monster);
+      }
+      // change instances
+      let r = e.detail.logs[i].match(instanceChangedRegex[lang] || instanceChangedRegex['en']);
+      if (r)
+        this.ClearTargetMonsters();
+      // removing new combatant
+      r = e.detail.logs[i].match(/19:(.*) was defeated by/);
+      if (r)
+        this.RemoveMonster(r[1]);
+    }
+  }
+
+  OnPlayerChange(e) {
+    this.playerPos['x'] = e.detail.pos.x;
+    this.playerPos['y'] = -e.detail.pos.y;
```

Comment:
I still think the negative y is confusing.  Is it possible to fix the math to operate on the real y coordinate?

Reply:
y is negative because y position of in-game map is upside down, like
![a.jpg](https://i.loli.net/2019/12/02/zbXJsi6MxoZ1aUA.jpg)
And negative y can transpose it to a standard [Cartesian Coordinate System](https://en.wikipedia.org/wiki/Cartesian_coordinate_system), which will make the math system more "standard".
Yes, we can fix the math, but that's only different in when you negative y coordinate (in the formula, pre-processing, etc).
Since `this.playerPos` is never shown to user so we can treat this as a variable used in the math system.

Reply:
I guess I don't think of that as upside down.  They're both valid cartesian coordinates, in my opinion.  The code is just confusing to me because sometimes y is negated (players) and sometimes it's not (monsters).

If you have x,y with y = down, then you can find the angle of the vector d = (monster x - player x, monster y - player y), with this equation: `let degrees = Math.atan2(d.y, d.x) / Math.PI * 180;`

e.g. a delta of (10, 0) is 0 degrees, a delta of (0, 10) is 90 degrees (pointing down).

I think this simplifies the rotation math some too, rather than using acos.

Reply:
Yes, this is kinda confusing because the `pos_y` of monster is not negatived, it represents the true position of monsters, and thus the position shown in the overlay won't be affected.
When doing the math, what we actually use is the `pos` of monsters, it's negatived as I said, to be used in the math.
But yes, I should make the rotation calculating simple and clear, I'll fix this after.